File: tests/fixtures/csharp/AuthService.cs
using System;
using System.Collections.Generic;

namespace CodemarkExample.Auth
{
    // Target: bookmark a record
    public record Claims(string Subject, DateTime Expiry, List<string> Roles);

    // Target: bookmark an enum
    public enum AuthError
    {
        InvalidToken,
        Expired,
        InsufficientPermissions
    }

    // Target: bookmark an interface
    public interface IAuthProvider
    {
        Claims ValidateToken(string token);
        string RefreshToken(string token);
    }

    // Target: bookmark a class
    public class AuthService : IAuthProvider
    {
        private readonly string _secret;
        private readonly string _issuer;
        private readonly Dictionary<string, Claims> _tokenCache = new();

        public AuthService(string secret, string issuer = "codemark")
        {
            _secret = secret;
            _issuer = issuer;
        }

        // Target: bookmark a method
        public Claims ValidateToken(string token)
        {
            if (_tokenCache.TryGetValue(token, out var cached))
            {
                if (cached.Expiry <= DateTime.UtcNow)
                {
                    _tokenCache.Remove(token);
                    throw new InvalidOperationException("Token expired");
                }
                return cached;
            }

            var claims = Decode(token);
            _tokenCache[token] = claims;
            return claims;
        }

        public string RefreshToken(string token)
        {
            var claims = Decode(token);
            return Encode(claims);
        }

        // Target: bookmark a private method
        private Claims Decode(string token)
        {
            if (string.IsNullOrEmpty(token))
            {
                throw new ArgumentException("Invalid token");
            }
            return new Claims("user-1", DateTime.UtcNow.AddHours(1), new List<string> { "admin" });
        }

        private string Encode(Claims claims)
        {
            return $"{_issuer}:{claims.Subject}:{_secret}";
        }

        // Target: bookmark a method with complex params
        public void CheckPermission(Claims claims, string required, string resource)
        {
            if (!claims.Roles.Contains(required))
            {
                throw new UnauthorizedAccessException(
                    $"Insufficient permissions: {required} on {resource}");
            }
        }

        // Target: bookmark a static factory
        public static AuthService CreateDefault()
        {
            return new AuthService("default-secret");
        }
    }
}

File: tests/fixtures/swift/auth_service.swift
import Foundation

// MARK: - Models

struct Claims: Codable {
    let subject: String
    let expiry: Date
    let roles: [String]
}

enum AuthError: Error {
    case invalidToken
    case expired
    case insufficientPermissions(required: String)
}

// MARK: - Protocol

protocol AuthProvider {
    func validateToken(_ token: String) async throws -> Claims
    func refreshToken(_ token: String) async throws -> String
}

// MARK: - Implementation

class AuthService: AuthProvider {

    private let secret: String
    private let issuer: String
    private var tokenCache: [String: Claims] = [:]

    init(secret: String, issuer: String = "codemark") {
        self.secret = secret
        self.issuer = issuer
    }

    // Target: bookmark this function, rename it, and verify resolution still works.
    func validateToken(_ token: String) async throws -> Claims {
        if let cached = tokenCache[token] {
            guard cached.expiry > Date() else {
                tokenCache.removeValue(forKey: token)
                throw AuthError.expired
            }
            return cached
        }

        let claims = try decode(token)

        guard claims.expiry > Date() else {
            throw AuthError.expired
        }

        tokenCache[token] = claims
        return claims
    }

    // Target: bookmark this, then extract it to a separate file and verify cross-file resolution.
    func refreshToken(_ token: String) async throws -> String {
        let claims = try decode(token)

        guard claims.expiry > Date().addingTimeInterval(-3600) else {
            throw AuthError.expired
        }

        let newClaims = Claims(
            subject: claims.subject,
            expiry: Date().addingTimeInterval(3600),
            roles: claims.roles
        )

        return try encode(newClaims)
    }

    // Target: bookmark a private method nested inside a class.
    private func decode(_ token: String) throws -> Claims {
        guard let data = Data(base64Encoded: token) else {
            throw AuthError.invalidToken
        }
        return try JSONDecoder().decode(Claims.self, from: data)
    }

    private func encode(_ claims: Claims) throws -> String {
        let data = try JSONEncoder().encode(claims)
        return data.base64EncodedString()
    }

    // Target: bookmark a method with complex parameters for disambiguation testing.
    func checkPermission(_ claims: Claims, required: String, resource: String) throws -> Bool {
        guard claims.roles.contains(required) else {
            throw AuthError.insufficientPermissions(required: required)
        }
        return true
    }
}

// MARK: - Extension

extension AuthService {
    // Target: bookmark a method inside an extension.
    func invalidateCache() {
        tokenCache.removeAll()
    }

    func cacheSize() -> Int {
        return tokenCache.count
    }
}

// MARK: - Free function (top-level)

// Target: bookmark a top-level function (no class/struct parent).
func createDefaultAuthService() -> AuthService {
    return AuthService(secret: "default-secret")
}
